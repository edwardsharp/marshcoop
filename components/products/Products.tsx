import React from 'react'
import useSWR from 'swr'

import { getProducts } from '../../services/products'

export default function Products() {
  const { data: products, error } = useSWR('get_products', () =>
    getProducts(10)
  )

  if (error) return <div>failed to load</div>
  if (!products) return <div>loading . . .</div>

  return (
    <>
      <h2>{products.length} Products</h2>
      <table>
        <thead>
          {products[0] && (
            <tr>
              {Object.keys(products[0]).map((k) => (
                <td key={`${k}`}>
                  <dd>{k}</dd>
                </td>
              ))}
            </tr>
          )}
        </thead>
        <tbody>
          {products.map((product) => (
            <tr key={product.id}>
              {Object.values(product).map((v) => (
                <td key={`${product.id}${v}`}>
                  <dd>{v}</dd>
                </td>
              ))}
            </tr>
          ))}
        </tbody>
      </table>
    </>
  )
}
