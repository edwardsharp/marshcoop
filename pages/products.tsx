import type { NextPage } from 'next'
import Head from 'next/head'
import ProductsList from '../components/products/ProductsList'

const ProductsPage: NextPage = () => {
  return (
    <>
      <Head>
        <title>MARSH COOP</title>
        <meta name="description" content="MARSH COOP" />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <ProductsList />
    </>
  )
}

export default ProductsPage
