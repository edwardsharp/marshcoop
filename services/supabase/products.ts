import { SupabaseClient } from '@supabase/supabase-js'
import { definitions } from '../../types/supabase'
import { logEvent } from './events'
import { supabase } from './supabase'

export type Product = definitions['products']

export async function getProducts(limit = 1000) {
  const {
    data: products,
    error,
    count,
  } = await supabase
    .from<Product>('products')
    .select('*', { count: 'exact' })
    .limit(limit)

  if (error) throw new Error(error.message)
  return products
}

export async function getProductsCount() {
  const { error, count } = await supabase
    .from<Product>('products')
    .select('*', { count: 'exact' })

  if (error) throw new Error(error.message)
  return count
}

export async function getCategories(): Promise<{ [index: string]: string }> {
  const { data, error } = await supabase.rpc('distinct_product_categories')

  if (error) {
    throw new Error(error.message)
  }

  // #TODO: rework this to just reduce to a string array of .category items
  return data?.reduce((acc, row) => {
    acc[row.category] = row.category
    return acc
  }, {})
}

export async function getSubCategories(category: string) {
  const { data, error } = await supabase.rpc(
    'distinct_product_sub_categories',
    { category }
  )

  if (error) {
    throw new Error(error.message)
  }

  // #TODO: rework this to just reduce to a string array of .category items
  return data?.reduce((acc, row) => {
    acc[row.category] = row.category
    return acc
  }, {})
}

// all the functions below are known to need non-anon privileges, so client prop is passed in.
// there might be a better way; this is sort of ad-hoc DI :internet-shrugz:
export async function upsertProducts(props: {
  products: Product | Product[]
  client?: SupabaseClient
}) {
  const { client, products } = props
  const c = client ? client : supabase

  // #TODO: this should probably go thru each product and check if it already exists.
  // and then only update a few props (like count_on_hand, updated_at, maybe price?)
  // avoids ambigous changes like unit_type (which might be hard to determine from square data?)
  // ...maybe? :hmm: need to think more about this.
  const { data, error, count } = await c.from('products').upsert(products)

  if (error) {
    await logEvent({
      tag: 'upsertProducts',
      message: `upsertProducts error: ${error.message})`,
      level: 'error',
    })
  }
  await logEvent({
    tag: 'upsertProducts',
    message: `upsertProducts result count: ${count})`,
    level: 'debug',
    data: JSON.stringify({ data, error, count }),
  })
  return { data, error, count }
}

export async function createOrUpdateProducts(props: {
  products: Product[]
  client?: SupabaseClient
}) {
  const { client, products } = props
  const c = client ? client : supabase

  let createdProducts = 0
  let updatedProducts = 0

  for await (const product of products) {
    const { count: productCount } = await c
      .from('products')
      .select('id', { count: 'exact' })
      .eq('id', product.id)
    if (productCount) {
      // product exists, just update a couple properties
      const { count_on_hand, sq_variation_id } = product
      const { error } = await c
        .from<Product>('products')
        .update({ count_on_hand, sq_variation_id }, { returning: 'minimal' })
        .eq('id', product.id)
      if (error) {
        await logEvent({
          tag: 'createOrUpdateProducts',
          message: `update error: ${error.message})`,
          level: 'error',
          data: JSON.stringify({ error }),
        })
      } else {
        updatedProducts += 1
      }
    } else {
      const { error } = await c
        .from<Product>('products')
        .insert(product, { returning: 'minimal' })
      if (error) {
        await logEvent({
          tag: 'createOrUpdateProducts',
          message: `insert error: ${error.message})`,
          level: 'error',
          data: JSON.stringify({ error }),
        })
      } else {
        createdProducts += 1
      }
    }
  }

  await logEvent({
    tag: 'createOrUpdateProducts',
    message: `createOrUpdateProducts result createdProducts: ${createdProducts}, updatedProducts: ${updatedProducts}`,
  })
}

export async function updateProductCountOnHand(props: {
  variation_id: string
  count_on_hand: number
  client?: SupabaseClient
}) {
  const { variation_id, count_on_hand, client } = props
  const c = client ? client : supabase
  const { data, error } = await c
    .from<Product>('products')
    .update({ count_on_hand })
    .match({ variation_id })

  //   if (error) throw new Error(error.message)
  if (error) {
    await logEvent({
      tag: 'updateProductCountOnHand',
      message: `updateProductCountOnHand caught error: ${error.message})`,
      level: 'error',
    })
  }
  return data
}
