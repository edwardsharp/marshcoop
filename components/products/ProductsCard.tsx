import React from 'react'
import useSWR from 'swr'

import { getProductsCount } from '../../services/supabase/products'

export default function ProductsCard() {
  const { data: count, error } = useSWR('get_products_card', getProductsCount)
  console.log('count, error', count, error)
  if (error) return <div>failed to load</div>
  if (!count) return <div>loading...</div>

  return (
    <>
      <h2>Back Catalog</h2>
      <p>{count} products</p>
    </>
  )
}
