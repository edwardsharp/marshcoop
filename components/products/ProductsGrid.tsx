import React, { useState } from 'react'
import useSWR from 'swr'

import { getProducts, Product } from '../../services/supabase/products'
import { supabase } from '../../services/supabase/supabase'
import styles from '../../styles/Grid.module.css'

function formatPrice(product: Product) {
  if (!product.u_price || !product.ws_price) {
    return ''
  }
  const u_price = parseFloat(`${product.u_price}`).toFixed(2)
  if (product.ws_price === product.u_price) {
    return (
      <>
        <h2>${u_price}</h2>
        <div>
          <i>
            {product.pk && product.pk > 1 ? `${product.pk} pk` : ''}{' '}
            {product.size}
          </i>
        </div>
      </>
    )
  }
  const ws_price = parseFloat(`${product.ws_price}`).toFixed(2)
  return (
    <>
      <h2>${ws_price}</h2>{' '}
      <div style={{ display: 'flex', flexDirection: 'column' }}>
        <i>
          {product.pk && product.pk > 1 ? `${product.pk} pk` : ''}{' '}
          {product.size}
        </i>
        <i>${u_price} ea.</i>
      </div>
    </>
  )
}

interface ProductsGrouped {
  [index: string]: { [index: string]: Product[] }
}

interface ProductsData {
  grouped: ProductsGrouped
  cats: string[]
  count: number
}

const Grid = (props: { products: ProductsData; error: any }) => {
  const { products, error } = props
  if (error) return <div>failed to load</div>
  if (!products) return <div>loading . . .</div>

  return (
    <>
      {products.cats.map((cat) => (
        <div key={cat}>
          {products.grouped &&
            products.grouped[cat] &&
            Object.keys(products.grouped[cat]).map((sub_cat) => (
              <>
                <div key={sub_cat} className={styles.grid}>
                  <div className={styles.sticky}>
                    <h1>{cat}</h1>
                    <h2>
                      {sub_cat} (
                      {products.grouped &&
                        products.grouped[cat][sub_cat].length}
                      ){' '}
                    </h2>
                  </div>
                  {products.grouped &&
                    products.grouped[cat][sub_cat].map((product) => (
                      <div className={styles.card} key={product.id}>
                        <p>{product.description}</p>
                        <div>{formatPrice(product)}</div>
                        <small>{product.name}</small>
                      </div>
                    ))}
                </div>
              </>
            ))}
        </div>
      ))}
    </>
  )
}

export default function ProductGrid() {
  const [searchQ, setSearchQ] = useState('')
  const { data: products, error } = useSWR(
    { key: 'get_product', searchQ },
    async ({ searchQ: q }) => {
      let query = supabase.from('products').select('*', { count: 'exact' })
      if (q) {
        query = query.textSearch('fts', q, {
          type: 'websearch',
          config: 'english',
        })
      }
      const { data: products, error, count } = await query

      let cats: string[] = []
      const grouped = products?.reduce((acc, product) => {
        const cat = product.category || 'no category'
        if (!acc[cat]) {
          acc[cat] = {}
        }
        if (!cats.includes(cat)) {
          cats.push(cat)
        }
        const sub_cat = product.sub_category || ''
        if (!acc[cat][sub_cat]) {
          acc[cat][sub_cat] = []
        }
        acc[cat][sub_cat].push(product)

        return acc
      }, {} as ProductsGrouped)

      return { grouped, cats, count: count || 0 }
    }
  )

  return (
    <>
      <div
        style={{
          height: 60,
          position: 'sticky',
          top: 0,
          background: 'white',
          zIndex: 1,
          margin: '0 1em',
          display: 'flex',
          justifyContent: 'space-between',
          alignItems: 'center',
        }}
      >
        <div>{products && products.count} products</div>
        <div>
          <input
            type="search"
            placeholder="Search"
            value={searchQ}
            onChange={(e) => setSearchQ(e.target.value)}
          />

          {/* <button>:)</button> */}
        </div>
      </div>

      {products && <Grid products={products} error={error} />}
    </>
  )
}
