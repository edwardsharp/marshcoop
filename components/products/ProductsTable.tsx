import React from 'react'
import useSWR from 'swr'

import { getProducts, Product } from '../../services/supabase/products'

export default function ProductsTable() {
  const { data: products, error } = useSWR('get_products', () =>
    getProducts(1000)
  )

  const columns: Array<keyof Product> = [
    'category',
    'sub_category',
    'description',
    'pk',
    'size',
    'unit_type',
    'ws_price',
    'u_price',
    'codes',
    'count_on_hand',
  ]

  if (error) return <div>failed to load</div>
  if (!products) return <div>loading . . .</div>

  return (
    <>
      <div
        style={{
          height: 60,
          position: 'sticky',
          top: 0,
          background: 'white',
          zIndex: 1,
          margin: '0 1em',
          display: 'flex',
          justifyContent: 'space-between',
          alignItems: 'center',
        }}
      >
        <div>{products.length} Products</div>
        <div>
          <input type="search" placeholder="Search" />

          <button>:)</button>
        </div>
      </div>
      <table style={{ margin: '0 1em' }}>
        <thead>
          <tr
            style={{
              position: 'sticky',
              top: 50,
              background: 'white',
              zIndex: 1,
            }}
          >
            {columns.map((col) => (
              <td key={`${col}`}>{col}</td>
            ))}
            <td>&nbsp;</td>
          </tr>
        </thead>
        <tbody>
          {products.map((product) => (
            <tr key={product.id}>
              {columns.map((col) => (
                <td key={`${product.id}${col}`}>{product[col]}</td>
              ))}
              <td>
                <button>+</button>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </>
  )
}
