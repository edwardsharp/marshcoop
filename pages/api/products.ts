import type { NextApiRequest, NextApiResponse } from 'next'

import { defaultCorsMiddleware } from '../../lib/cors-middleware'
import { getProducts, Product } from '../../services/supabase/products'

type Data = {
  data: Product[] | null
  page: number
  totalCount: number
}

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse<Data>
) {
  await defaultCorsMiddleware(req, res)

  const products = await getProducts()

  res.status(200).json({ data: products, page: 0, totalCount: 1000 })
}
