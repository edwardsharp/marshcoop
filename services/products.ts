import { PostgrestError } from '@supabase/supabase-js'
import { definitions } from '../types/supabase'
import { supabase } from './supabase'

export type Product = definitions['products']

export async function getProducts(limit = 1000) {
  return await supabase.from<Product>('products').select().limit(limit)
}

export function getProductsSuspense(limit = 1000) {
  let status = 'pending'
  let result: Product[] //| PostgrestError

  const suspender = supabase
    .from<Product>('products')
    .select()
    .limit(limit)
    .then((products) => {
      if (products.error) {
        status = 'error'
        result = []
      } else {
        status = 'success'
        result = products.data
      }
    })

  return {
    read() {
      if (status === 'pending') {
        throw suspender
      } else if (status === 'error') {
        throw result
      } else if (status === 'success') {
        return result
      }
    },
  }
}
