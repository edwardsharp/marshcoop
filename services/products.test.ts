import { getProducts } from './products'

describe('products service', () => {
  test('getProducts matches snapshot', async () => {
    const products = await getProducts(10)
    expect(products).toMatchSnapshot()
  })

  // hmm mock getProducts in the following?
  test('getProducts with no aguments', async () => {
    const products = await getProducts()
    expect(products && products.length).toBeGreaterThanOrEqual(1)
  })

  test('getProducts throws', async () => {
    // note: jest needs whatever fn that throws to be wrapped in a fn, so the lambda here:
    await expect(async () => await getProducts(-1)).rejects.toThrow()
  })
})
