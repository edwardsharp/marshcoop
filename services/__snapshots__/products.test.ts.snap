// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`products service getProducts matches snapshot 1`] = `
Array [
  Object {
    "category": "ALBERT'S FRESH",
    "codes": "g",
    "count_on_hand": null,
    "createdAt": "2021-01-02T00:59:46.649+00:00",
    "description": "Butter, Salted, Irish, Gluten-Free, Kerrygold",
    "id": 63494,
    "import_tag": "albertz",
    "name": "KERRYGOLD",
    "no_backorder": false,
    "pk": 20,
    "size": "8 oz",
    "sub_category": "DAIRY",
    "u_price": 3.83,
    "u_price_cost": 3.83,
    "unf": "86037",
    "unit_type": "CS",
    "upc_code": "6-16973-71180-7",
    "updatedAt": "2021-01-02T00:59:46.649+00:00",
    "vendor": "albert's fresh",
    "ws_price": 76.51,
    "ws_price_cost": 76.51,
  },
  Object {
    "category": "ALBERT'S FRESH",
    "codes": "g",
    "count_on_hand": null,
    "createdAt": "2021-01-02T00:59:46.649+00:00",
    "description": "Butter, Unsalted, Irish, Gluten-Free, Kerrygold",
    "id": 63495,
    "import_tag": "albertz",
    "name": "KERRYGOLD",
    "no_backorder": false,
    "pk": 20,
    "size": "8 oz",
    "sub_category": "DAIRY",
    "u_price": 3.83,
    "u_price_cost": 3.83,
    "unf": "84829",
    "unit_type": "CS",
    "upc_code": "6-16973-71115-9",
    "updatedAt": "2021-01-02T00:59:46.649+00:00",
    "vendor": "albert's fresh",
    "ws_price": 76.51,
    "ws_price_cost": 76.51,
  },
  Object {
    "category": "ALBERT'S FRESH",
    "codes": "2",
    "count_on_hand": null,
    "createdAt": "2021-01-02T00:59:46.649+00:00",
    "description": "Cheese, Chevre Plain Log, Amaltheia",
    "id": 63499,
    "import_tag": "albertz",
    "name": "AMALTHEIA",
    "no_backorder": false,
    "pk": 12,
    "size": "4 oz",
    "sub_category": "DAIRY",
    "u_price": 3.67,
    "u_price_cost": 3.67,
    "unf": "88140",
    "unit_type": "CS",
    "upc_code": "6-16973-11115-7",
    "updatedAt": "2021-01-02T00:59:46.649+00:00",
    "vendor": "albert's fresh",
    "ws_price": 44,
    "ws_price_cost": 44,
  },
  Object {
    "category": "ALBERT'S FRESH",
    "codes": "2",
    "count_on_hand": null,
    "createdAt": "2021-01-02T00:59:46.649+00:00",
    "description": "Cheese, Chevre, Roasted Garlic & Chive, Amaltheia",
    "id": 63500,
    "import_tag": "albertz",
    "name": "AMALTHEIA",
    "no_backorder": false,
    "pk": 12,
    "size": "4 oz",
    "sub_category": "DAIRY",
    "u_price": 3.73,
    "u_price_cost": 3.73,
    "unf": "84830",
    "unit_type": "CS",
    "upc_code": "6-16973-71116-6",
    "updatedAt": "2021-01-02T00:59:46.649+00:00",
    "vendor": "albert's fresh",
    "ws_price": 44.7,
    "ws_price_cost": 44.7,
  },
  Object {
    "category": "ALBERT'S FRESH",
    "codes": "2",
    "count_on_hand": null,
    "createdAt": "2021-01-02T00:59:46.649+00:00",
    "description": "Cheese, Parm Regg, Ambrosi",
    "id": 63501,
    "import_tag": "albertz",
    "name": "AMBROSI",
    "no_backorder": false,
    "pk": 10,
    "size": "7 oz",
    "sub_category": "DAIRY",
    "u_price": 9.45,
    "u_price_cost": 9.45,
    "unf": "81437",
    "unit_type": "CS",
    "upc_code": "6-16973-79029-1",
    "updatedAt": "2021-01-02T00:59:46.649+00:00",
    "vendor": "albert's fresh",
    "ws_price": 94.51,
    "ws_price_cost": 94.51,
  },
  Object {
    "category": "ALBERT'S FRESH",
    "codes": "2",
    "count_on_hand": null,
    "createdAt": "2021-01-02T00:59:46.649+00:00",
    "description": "Cheese, Parm Regg Flakes, Ambrosi",
    "id": 63502,
    "import_tag": "albertz",
    "name": "AMBROSI",
    "no_backorder": false,
    "pk": 6,
    "size": "2.82 oz",
    "sub_category": "DAIRY",
    "u_price": 4.6,
    "u_price_cost": 4.6,
    "unf": "53419",
    "unit_type": "CS",
    "upc_code": "6-16973-76020-1",
    "updatedAt": "2021-01-02T00:59:46.649+00:00",
    "vendor": "albert's fresh",
    "ws_price": 27.62,
    "ws_price_cost": 27.62,
  },
  Object {
    "category": "ALBERT'S FRESH",
    "codes": "2",
    "count_on_hand": null,
    "createdAt": "2021-01-02T00:59:46.649+00:00",
    "description": "Cheese, Parm Regg Grated, Ambrosi",
    "id": 63503,
    "import_tag": "albertz",
    "name": "AMBROSI",
    "no_backorder": false,
    "pk": 6,
    "size": "2.82 oz",
    "sub_category": "DAIRY",
    "u_price": 4.41,
    "u_price_cost": 4.41,
    "unf": "53421",
    "unit_type": "CS",
    "upc_code": "6-16973-78300-2",
    "updatedAt": "2021-01-02T00:59:46.649+00:00",
    "vendor": "albert's fresh",
    "ws_price": 26.43,
    "ws_price_cost": 26.43,
  },
  Object {
    "category": "ALBERT'S FRESH",
    "codes": "n",
    "count_on_hand": null,
    "createdAt": "2021-01-02T00:59:46.649+00:00",
    "description": "Cheese, Cheddar, Vintage, 24 Month Aged, Grassfed, Anchor",
    "id": 63504,
    "import_tag": "albertz",
    "name": "ANCHOR",
    "no_backorder": false,
    "pk": 12,
    "size": "7 oz",
    "sub_category": "DAIRY",
    "u_price": 3.61,
    "u_price_cost": 3.61,
    "unf": "53418",
    "unit_type": "CS",
    "upc_code": "6-16973-76010-2",
    "updatedAt": "2021-01-02T00:59:46.649+00:00",
    "vendor": "albert's fresh",
    "ws_price": 43.29,
    "ws_price_cost": 43.29,
  },
  Object {
    "category": "ALBERT'S FRESH",
    "codes": "n",
    "count_on_hand": null,
    "createdAt": "2021-01-02T00:59:46.649+00:00",
    "description": "Cheese, Egmont (Gouda Style), Grassfed, Anchor",
    "id": 63505,
    "import_tag": "albertz",
    "name": "ANCHOR",
    "no_backorder": false,
    "pk": 12,
    "size": "7 oz",
    "sub_category": "DAIRY",
    "u_price": 2.8,
    "u_price_cost": 2.8,
    "unf": "61776",
    "unit_type": "CS",
    "upc_code": "6-16973-71134-0",
    "updatedAt": "2021-01-02T00:59:46.649+00:00",
    "vendor": "albert's fresh",
    "ws_price": 33.57,
    "ws_price_cost": 33.57,
  },
  Object {
    "category": "ALBERT'S FRESH",
    "codes": "n",
    "count_on_hand": null,
    "createdAt": "2021-01-02T00:59:46.649+00:00",
    "description": "Cheese, Sharp, New Zealand, Grassfed, Anchor",
    "id": 63506,
    "import_tag": "albertz",
    "name": "ANCHOR",
    "no_backorder": false,
    "pk": 12,
    "size": "7 oz",
    "sub_category": "DAIRY",
    "u_price": 2.8,
    "u_price_cost": 2.8,
    "unf": "69203",
    "unit_type": "CS",
    "upc_code": "6-16973-70164-8",
    "updatedAt": "2021-01-02T00:59:46.649+00:00",
    "vendor": "albert's fresh",
    "ws_price": 33.57,
    "ws_price_cost": 33.57,
  },
]
`;
